crates/hir-ty/src/infer/unify.rs

```diff
@@ -31,6 +32,72 @@ impl InferenceContext<'_> {
     {
         self.table.canonicalize(t)
     }
+
+    pub(super) fn clauses_for_self_ty(
+        &mut self,
+        self_ty: InferenceVar,
+    ) -> SmallVec<[WhereClause; 4]> {
+        self.table.resolve_obligations_as_possible();
+
+        let root = self.table.var_unification_table.inference_var_root(self_ty);
+        let pending_obligations = mem::take(&mut self.table.pending_obligations);
+        let obligations = pending_obligations
+            .iter()
+            .filter_map(|obligation| match obligation.value.value.goal.data(Interner) {
```

Comment:
Could make use of https://doc.rust-lang.org/std/vec/struct.Vec.html#method.retain I think?

Reply:
Since we are not mutating `obligations` - it should be returned to its original place `pending_obligations` - I think that to use `Vec::retain()`, we should clone the whole `obligations`